Add spooling protocol to client REST API reference (#30549)
trinodb/trino: docs/src/main/sphinx/develop/client-protocol.md

provides a hint about the query state. It cannot be used to tell if the query
is finished.

(spooling-protocol-flow)=
## Spooling protocol

By default a client receives results with the direct protocol: the `data` field
of each `QueryResults` document holds the rows inline. A client can instead opt
in to the [spooling protocol](protocol-spooling), in which larger results are
staged to object storage and the client fetches them out of band. The spooling
protocol must be [enabled and configured on the server](protocol-spooling); the
client requests it per query.

### Requesting the spooling protocol

> A client opts in by sending the `X-Trino-Query-Data-Encoding` request header on

Reviewer: Do we want to add here or somewhere how a user does that on the client or SQL layer? Or link to wherever we have documented that already.

Author: Added a link to the client documentation rather than duplicating it here. Happy to inline a short version if you think it reads better.

  that must be sent with the retrieval and acknowledgement requests. Both `uri`
  and `ackUri` are opaque absolute URIs that may point at different hosts, so a
  client must use each as given rather than parsing it or assuming a common host.

Both types carry a `metadata` object describing the segment. A client decodes
segments in order, using the negotiated encoding, to reconstruct the result rows.

:::{list-table} Segment metadata attributes
:widths: 25, 15, 45
:header-rows: 1

* - Attribute
  - Type
  - Description
* - `rowOffset`
  - number
  - Absolute offset of the segment's first row within the result set.
* - `rowsCount`
  - number
  - Number of rows in the segment.
* - `segmentSize`
  - number
  - Size of the segment in bytes, as stored or transferred.
* - `uncompressedSize`
  - number
  - Size of the segment in bytes after decompression. Present only for compressed
    encodings.
* - `expiresAt`
  - string
  - Time after which a spooled segment may no longer be retrievable from storage.
:::

Early pages of a large result are often returned as `inline` segments even when
spooling is requested, because small results do not benefit from spooling. The

> server controls this with `protocol.spooling.inlining.enabled` (default enabled).

Reviewer: ```suggestion
server controls this with `protocol.spooling.inlining.enabled`, which defaults to `true`.
```

### Retrieving a spooled segment

How a client fetches the data for a `spooled` segment depends on the server's
`protocol.spooling.retrieval-mode`. `STORAGE` is the default, and the other modes
change where the data is read from:

:::{list-table} Segment retrieval modes
:widths: 25, 45, 30
:header-rows: 1

* - Mode

>   - How the client retrieves segment data

Reviewer: ```suggestion
  - Description
```

>   - Client needs network access to

Reviewer: ```suggestion
  - Required client network access
```

* - `STORAGE`
  - The segment `uri` is a pre-signed object storage URI. The client reads the
    data directly from storage.
  - Object storage
* - `COORDINATOR_STORAGE_REDIRECT`
  - The `uri` points at the coordinator, which responds with a redirect to a
    pre-signed object storage URI.
  - Coordinator and object storage
* - `COORDINATOR_PROXY`
  - The `uri` points at the coordinator, which reads from storage and streams
    the data back itself.
  - Coordinator only
* - `WORKER_PROXY`
  - The `uri` points at the coordinator, which redirects to a worker that streams
    the data back.
  - Coordinator and workers
:::

In `STORAGE` and `COORDINATOR_STORAGE_REDIRECT`, the client reads from object
storage directly, so it must have network access to the configured storage
location. In `COORDINATOR_PROXY` and `WORKER_PROXY`, the cluster reads from
storage on the client's behalf, so the client does not need storage access.

A `spooled` segment may include a `headers` map. These headers are opaque to the

> client: it must send them unchanged on the retrieval request, and on the

Reviewer: ```suggestion
client. It must send them unchanged on the retrieval request, and on the
```

acknowledgement request, without interpreting or transforming them.

### Acknowledging a segment

> After a client has read a `spooled` segment, it should acknowledge it with a

Reviewer: ```suggestion
After a client has read a `spooled` segment, it must acknowledge processing with a
```

although "must" is a bit strong and it works without .. but not well really since it relies on TTL cleanup

Author: Kept "should" rather than "must" per your note that it works (via TTL cleanup) without acknowledgement.

Reviewer 2: Proper client implementation must call ackUri. It's up to the server whether to ignore it. Manfred is right on that.

Author: You're right — changed to "must" and reframed so acknowledgement is a required client step, with whether the server acts on it left as a server-side decision.

`GET` to the segment's `ackUri`, sending the same `headers` that came with the
segment. Acknowledgement lets the server remove the segment from object storage
promptly rather than waiting for TTL-based cleanup.

> The `ackUri`, like the segment `uri`, is opaque: a client issues the `GET` to it

Reviewer: ```suggestion
The `ackUri`, like the segment `uri`, is opaque. A client issues the `GET` to it
```

exactly as provided. An intermediary between the client and Trino, such as a load

> balancer or gateway, must forward requests to these URIs so that segment

Reviewer: ```suggestion
balancer, router, or Trino Gateway, must forward requests to these URIs so that segment
```

retrieval and acknowledgement reach their destination.

> Acknowledgement is not required for correctness. Segments that are never

Reviewer: Reword this .. too much "acknowledge"

      "data": "<base64-encoded encoded rows>",
      "metadata": { "rowOffset": 0, "rowsCount": 10, "segmentSize": 1024 }
    },
    {
      "type": "spooled",
      "uri": "https://coordinator.example/v1/spooled/download/<id>",
      "ackUri": "https://coordinator.example/v1/spooled/ack/<id>",
      "metadata": { "rowOffset": 10, "rowsCount": 5000, "segmentSize": 4194304 },
      "headers": {
        "x-amz-server-side-encryption-customer-algorithm": ["AES256"],
        "x-amz-server-side-encryption-customer-key": ["<ephemeral key>"],
        "x-amz-server-side-encryption-customer-key-MD5": ["<md5>"]
      }
    }
  ]
}
```

A segment is one of two types, distinguished by the `type` field:

- `inline` — the rows are embedded directly in the `data` field of the segment,
  base64-encoded in the negotiated encoding. No further request is needed.
- `spooled` — the rows are staged on object storage. The segment carries a `uri`
  to retrieve the data, an `ackUri` to acknowledge it, and optional `headers`
  that must be sent with the retrieval and acknowledgement requests. The `uri` and
  `ackUri` may point at different hosts: depending on the retrieval mode the `uri`

>   can resolve to object storage while the `ackUri` always targets the coordinator.

Reviewer 2: this is not accurate. In the current implementation it points to ackUri but generally it's opaque so it can point to wherever in the future.

Author: Good point — reworded both this and the acknowledgement section to treat the URIs as opaque.

  A client must treat them as independent absolute URIs rather than assuming a
  common host.

Both types carry a `metadata` object describing the segment. A client decodes
segments in order, using the negotiated encoding, to reconstruct the result rows.

:::{list-table} Segment metadata attributes
:widths: 25, 15, 45
:header-rows: 1

* - Attribute
  - Type
  - Description
* - `rowOffset`
  - number

>   - Offset of the segment's first row within the whole result set.

Reviewer 2: Absolute offset of the segment's first row within the result set

Author: Done

change where the data is read from:

:::{list-table} Segment retrieval modes
:widths: 25, 45, 30
:header-rows: 1

* - Mode
  - How the client retrieves segment data
  - Client needs network access to
* - `STORAGE`
  - The segment `uri` is a pre-signed object storage URI. The client reads the
    data directly from storage.
  - Object storage
* - `COORDINATOR_STORAGE_REDIRECT`
  - The `uri` points at the coordinator, which responds with a redirect to a
    pre-signed object storage URI.
  - Coordinator and object storage
* - `COORDINATOR_PROXY`
  - The `uri` points at the coordinator, which reads from storage and streams
    the data back itself.
  - Coordinator only
* - `WORKER_PROXY`
  - The `uri` points at the coordinator, which redirects to a worker that streams
    the data back.
  - Coordinator and workers
:::

In `STORAGE` and `COORDINATOR_STORAGE_REDIRECT`, the client reads from object
storage directly, so it must have network access to the configured storage
location. In `COORDINATOR_PROXY` and `WORKER_PROXY`, the cluster reads from
storage on the client's behalf, so the client does not need storage access.

> When segment encryption is enabled (the default), a `spooled` segment includes

Reviewer 2: These headers are also opaque. Client is required to take them as-is and use it with GET requests. It doesn't need to understand its meaning or do any kind of a transformation. SSE-C is irrelevant in that context.

Author: Done — dropped the SSE-C detail and reworded to treat the headers as opaque. Also genericized the header names in the JSON example.

`headers` carrying the SSE-C headers for that segment. In the storage-facing
modes the client must send these headers on the retrieval request so that object
storage can decrypt the segment. The ephemeral encryption key is per-segment and
travels with the segment; the server-side `protocol.spooling.shared-secret-key`
is never sent to clients.

### Acknowledging a segment

After a client has read a `spooled` segment, it should acknowledge it with a
`GET` to the segment's `ackUri`, sending the same `headers` that came with the
segment. Acknowledgement lets the server remove the segment from object storage
promptly rather than waiting for TTL-based cleanup.

> The `ackUri` always points at the coordinator, regardless of the retrieval mode,

Reviewer 2: Inaccurate. This is true for the current implementation but generally ackUri is opaque

Author: Already addressed in the previous push — the acknowledgement section now treats ackUri as opaque too.
